### ISSUE TITLE
Abort tasks on backfill/GetSafeTime during Shutdown could call Wait on an un-waitable thread.

### ISSUE THREAD

**Issue by amitanandaiyer** [yugabyte/yugabyte-db#4681]
Jira Link: [DB-5947](https://yugabyte.atlassian.net/browse/DB-5947)
Ran into this while running `ybd --cxx-test tools_yb-admin-test_ent --gtest_filter AdminCliTest.TestExportImportIndexSnapshot --test_args --vmodule="backfill_index=3" --test_args --disable_index_backfill=false`

```
20736 Fatal failure details written to /net/dev-server-amitanand2/share/code/yugabyte-db/build/debug-gcc-dynamic-ninja/yb-test-logs/tests-tools__yb-admin-test_ent/AdminCliTest_TestExportImportIndexSnapshot.fatal_failure_details.20      20-06-04T18_54_27.pid14726.txt
20737 F20200604 18:54:27 ../../src/yb/util/thread_restrictions.cc:91] Check failed: LoadTLS()->wait_allowed Waiting is not allowed to be used on this thread to prevent server-wide latency aberrations and deadlocks. Thread 14736 (n      ame: "Master_reactor", category: "Master_reactor")
20738     @     0x7fb17ef70593  yb::LogFatalHandlerSink::send(int, char const*, char const*, int, tm const*, char const*, unsigned long) (src/yb/util/logging.cc:474)
20739     @     0x7fb17e149335
20740     @     0x7fb17e146799
20741     @     0x7fb17e149868
20742     @     0x7fb17effa12d  yb::ThreadRestrictions::AssertWaitAllowed() (src/yb/util/thread_restrictions.cc:91)
20743     @     0x7fb17eeec998  yb::CountDownLatch::WaitUntil(yb::MonoTime) const (src/yb/util/countdown_latch.cc:66)
20744     @     0x7fb17eeeca34  yb::CountDownLatch::WaitUntil(std::chrono::time_point<yb::CoarseMonoClock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) const (src/yb/util/countdown_latch.cc:59)
20745     @     0x7fb18abe172c  yb::master::SysCatalogTable::SyncWrite(yb::master::SysCatalogWriter*) (src/yb/master/sys_catalog.cc:627)
20746     @     0x7fb18aac7d74  yb::Status yb::master::SysCatalogTable::MutateItems<yb::master::TableInfo>(std::vector<yb::master::TableInfo*, std::allocator<yb::master::TableInfo*> > const&, yb::QLWriteRequestPB_QLStmtType const&      , long) (src/yb/master/sys_catalog-internal.h:126)
20747     @     0x7fb18aac8015
20748     @     0x7fb18aabbb1c  yb::master::MultiStageAlterTable::UpdateIndexPermission(yb::master::CatalogManager*, scoped_refptr<yb::master::TableInfo> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::alloca      tor<char> > const&, yb::IndexPermissions) (src/yb/master/backfill_index.cc:231)
20749     @     0x7fb18aabde69  yb::master::BackfillTable::AlterTableStateToAbort() (src/yb/master/backfill_index.cc:612)
20750     @     0x7fb18aac2baa  yb::master::BackfillTable::UpdateSafeTime(yb::Status const&, yb::HybridTime) (src/yb/master/backfill_index.cc:511)
20751     @     0x7fb18aac3936  yb::master::GetSafeTimeForTablet::UnregisterAsyncTaskCallback() (src/yb/master/backfill_index.cc:886)
20752     @     0x7fb18aa93423  yb::master::RetryingTSRpcTask::UnregisterAsyncTask() (src/yb/master/async_rpc_tasks.cc:350)
20753     @     0x7fb18aa935e5  yb::master::RetryingTSRpcTask::AbortAndReturnPrevState(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:205)
20754     @     0x7fb18aa9234f  yb::master::RetryingTSRpcTask::AbortTask(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:216)
20755     @     0x7fb18aa95a79  yb::master::RetryingTSRpcTask::RunDelayedTask(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:333)
20756     @     0x7fb18aaa1496
20757     @     0x7fb1806d1008  boost::function1<void, yb::Status const&>::operator()(yb::Status const&) const (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdparty/installed/uninstrumented/include/boost/functi      on/function_template.hpp:764)
20758     @     0x7fb1806d1008  yb::rpc::DelayedTask::AbortTask(yb::Status const&) (src/yb/rpc/reactor.cc:819)
20759     @     0x7fb1806d10b3  yb::rpc::DelayedTask::DoAbort(yb::Status const&) (src/yb/rpc/reactor.cc:828)
20760     @     0x7fb1806d0705  yb::rpc::ReactorTask::Abort(yb::Status const&) (src/yb/rpc/reactor.cc:735)
20761     @     0x7fb1806d49f3  yb::rpc::Reactor::ShutdownInternal() (src/yb/rpc/reactor.cc:235)
20762     @     0x7fb1806d4e2a  yb::rpc::Reactor::AsyncHandler(ev::async&, int) (src/yb/rpc/reactor.cc:351)
20763     @     0x7fb1806dbe6b  void ev::base<ev_async, ev::async>::method_thunk<yb::rpc::Reactor, &yb::rpc::Reactor::AsyncHandler>(ev_loop*, ev_async*, int) (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdpart      y/installed/common/include/ev++.h:479)
20764     @     0x7fb17cbd7b2a
20765     @     0x7fb17cbdbc39
20766     @     0x7fb1806cf62e  ev::loop_ref::run(int) (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdparty/installed/common/include/ev++.h:211)
20767     @     0x7fb1806cf62e  yb::rpc::Reactor::RunThread() (src/yb/rpc/reactor.cc:481)
20768     @     0x7fb1806d7ecf  void std::_Mem_fn_base<void (yb::rpc::Reactor::*)(), true>::operator()<, void>(yb::rpc::Reactor*) const (gcc/5.5.0_4/include/c++/5.5.0/functional:600)
20769     @     0x7fb1806d7ecf  void std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)>::__call<void, , 0ul>(tuple<>&&, std::_Index_tuple<0ul>) (gcc/5.5.0_4/include/c++/5.5.0/functional:1074)
20770     @     0x7fb1806d7ecf  void std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)>::operator()<, void>() (gcc/5.5.0_4/include/c++/5.5.0/functional:1133)
20771     @     0x7fb1806d7ecf  std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)> >::_M_invoke(std::_Any_data const&) (gcc/5.5.0_4/include/c++/5.5.0/functional:1871)
20772     @     0x7fb17eff64c9  std::function<void ()>::operator()() const (gcc/5.5.0_4/include/c++/5.5.0/functional:2267)
20773     @     0x7fb17eff64c9  yb::Thread::SuperviseThread(void*) (src/yb/util/thread.cc:744)
20774     @     0x7fb17a649693  start_thread (/tmp/glibc-20181130-26094-cs1x60/glibc-2.23/nptl/pthread_create.c:333)
20775     @     0x7fb17a38b41c  (unknown) (sysdeps/unix/sysv/linux/x86_64/clone.S:109)
20776     @ 0xffffffffffffffff
```

```
23371 W0604 18:38:05.524251 10760 thread.cc:600] Waited for 1.000s trying to join with Master_reactor (tid 10769)
23372 *** Check failure stack trace: ***
23373     @     0x7fd43295095e  DumpStackTraceAndExit (src/yb/util/logging.cc:166)
23374     @     0x7fd431b28c4c
23375     @     0x7fd431b2adbc
23376     @     0x7fd431b28799
23377     @     0x7fd431b2b868
23378     @     0x7fd4329dc12d  yb::ThreadRestrictions::AssertWaitAllowed() (src/yb/util/thread_restrictions.cc:91)
23379     @     0x7fd4328ce998  yb::CountDownLatch::WaitUntil(yb::MonoTime) const (src/yb/util/countdown_latch.cc:66)
23380     @     0x7fd4328cea34  yb::CountDownLatch::WaitUntil(std::chrono::time_point<yb::CoarseMonoClock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) const (src/yb/util/countdown_latch.cc:59)
23381     @     0x7fd43e5c372c  yb::master::SysCatalogTable::SyncWrite(yb::master::SysCatalogWriter*) (src/yb/master/sys_catalog.cc:627)
23382     @     0x7fd43e4a9d74  yb::Status yb::master::SysCatalogTable::MutateItems<yb::master::TableInfo>(std::vector<yb::master::TableInfo*, std::allocator<yb::master::TableInfo*> > const&, yb::QLWriteRequestPB_QLStmtType const&      , long) (src/yb/master/sys_catalog-internal.h:126)
23383     @     0x7fd43e4aa015
23384     @     0x7fd43e49db1c  yb::master::MultiStageAlterTable::UpdateIndexPermission(yb::master::CatalogManager*, scoped_refptr<yb::master::TableInfo> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::alloca      tor<char> > const&, yb::IndexPermissions) (src/yb/master/backfill_index.cc:231)
23385     @     0x7fd43e49fe69  yb::master::BackfillTable::AlterTableStateToAbort() (src/yb/master/backfill_index.cc:612)
23386     @     0x7fd43e4a4baa  yb::master::BackfillTable::UpdateSafeTime(yb::Status const&, yb::HybridTime) (src/yb/master/backfill_index.cc:511)
23387     @     0x7fd43e4a5936  yb::master::GetSafeTimeForTablet::UnregisterAsyncTaskCallback() (src/yb/master/backfill_index.cc:886)
23388     @     0x7fd43e475423  yb::master::RetryingTSRpcTask::UnregisterAsyncTask() (src/yb/master/async_rpc_tasks.cc:350)
23389     @     0x7fd43e4755e5  yb::master::RetryingTSRpcTask::AbortAndReturnPrevState(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:205)
23390     @     0x7fd43e47434f  yb::master::RetryingTSRpcTask::AbortTask(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:216)
23391     @     0x7fd43e477a79  yb::master::RetryingTSRpcTask::RunDelayedTask(yb::Status const&) (src/yb/master/async_rpc_tasks.cc:333)
23392     @     0x7fd43e483496
23393     @     0x7fd4340b3008  boost::function1<void, yb::Status const&>::operator()(yb::Status const&) const (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdparty/installed/uninstrumented/include/boost/functi      on/function_template.hpp:764)
23394     @     0x7fd4340b3008  yb::rpc::DelayedTask::AbortTask(yb::Status const&) (src/yb/rpc/reactor.cc:819)
23395     @     0x7fd4340b30b3  yb::rpc::DelayedTask::DoAbort(yb::Status const&) (src/yb/rpc/reactor.cc:828)
23396     @     0x7fd4340b2705  yb::rpc::ReactorTask::Abort(yb::Status const&) (src/yb/rpc/reactor.cc:735)
23397     @     0x7fd4340b69f3  yb::rpc::Reactor::ShutdownInternal() (src/yb/rpc/reactor.cc:235)
23398     @     0x7fd4340b6e2a  yb::rpc::Reactor::AsyncHandler(ev::async&, int) (src/yb/rpc/reactor.cc:351)
23399     @     0x7fd4340bde6b  void ev::base<ev_async, ev::async>::method_thunk<yb::rpc::Reactor, &yb::rpc::Reactor::AsyncHandler>(ev_loop*, ev_async*, int) (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdpart      y/installed/common/include/ev++.h:479)
23400     @     0x7fd4305b9b2a
23401     @     0x7fd4305bdc39
23402     @     0x7fd4340b162e  ev::loop_ref::run(int) (/n/jenkins/thirdparty/yugabyte-thirdparty-2020-04-20T01_46_16/thirdparty/installed/common/include/ev++.h:211)
23403     @     0x7fd4340b162e  yb::rpc::Reactor::RunThread() (src/yb/rpc/reactor.cc:481)
23404     @     0x7fd4340b9ecf  void std::_Mem_fn_base<void (yb::rpc::Reactor::*)(), true>::operator()<, void>(yb::rpc::Reactor*) const (gcc/5.5.0_4/include/c++/5.5.0/functional:600)
23405     @     0x7fd4340b9ecf  void std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)>::__call<void, , 0ul>(tuple<>&&, std::_Index_tuple<0ul>) (gcc/5.5.0_4/include/c++/5.5.0/functional:1074)
23406     @     0x7fd4340b9ecf  void std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)>::operator()<, void>() (gcc/5.5.0_4/include/c++/5.5.0/functional:1133)
23407     @     0x7fd4340b9ecf  std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (yb::rpc::Reactor::*)()> (yb::rpc::Reactor*)> >::_M_invoke(std::_Any_data const&) (gcc/5.5.0_4/include/c++/5.5.0/functional:1871)
23408     @     0x7fd4329d84c9  std::function<void ()>::operator()() const (gcc/5.5.0_4/include/c++/5.5.0/functional:2267)
23409     @     0x7fd4329d84c9  yb::Thread::SuperviseThread(void*) (src/yb/util/thread.cc:744)
23410     @     0x7fd42e02b693  start_thread (/tmp/glibc-20181130-26094-cs1x60/glibc-2.23/nptl/pthread_create.c:333)
23411     @     0x7fd42dd6d41c  (unknown) (sysdeps/unix/sysv/linux/x86_64/clone.S:109)
23412     @ 0xffffffffffffffff
```

[DB-5947]: https://yugabyte.atlassian.net/browse/DB-5947?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4681-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4681 (DB-5947)

Issue: "Abort tasks on backfill/GetSafeTime during Shutdown could call Wait on an un-waitable thread."
Opened 2020-06-04 by a maintainer from a debug-build test run (`AdminCliTest.TestExportImportIndexSnapshot`
with `--disable_index_backfill=false`); still open, no comments, labels `kind/bug area/docdb priority/medium`
(unchanged as of 2026-09-06). Kind: debug-build/test-infra crash (thread-restriction CHECK) during master shutdown.

## Verdict
NOT-REPRODUCED - the reported chain (reactor-thread abort of a scheduled backfill `GetSafeTimeForTablet`
retry -> `UpdateSafeTime` -> `AlterTableStateToAbort` -> synchronous sys-catalog `SyncWrite` on `Master_reactor`)
was exercised on the latest release of every in-support branch and the latest preview; on all five the callback
returns at a `kAborted` guard before any sys-catalog write, the master exits cleanly (exit 0, ~110 ms), and the
2020 defect code (unguarded callback) is only present in source before v2.7.2 / v2.15.3.0 (see provenance).
The literal FATAL cannot appear on official release images because `ThreadRestrictions::AssertWaitAllowed` is
compiled out under `NDEBUG` (`src/yb/util/thread_restrictions.h:37-39`, `:123-131`); the runtime evidence is
therefore a proxy for the path, backed by per-tag source inspection. Quick verdict: Fixed.

## Root cause
Historical defect (source at tag v2.2.0.0, July 2020, contemporaneous with the report):

- `src/yb/master/backfill_index.cc:1030-1049` (v2.2.0.0) - `GetSafeTimeForTablet::UnregisterAsyncTaskCallback()`
  unconditionally called `backfill_table_->UpdateSafeTime(status, safe_time)`, also when the task had been
  *aborted* (no response, `state() != kComplete` -> `InternalError` status).
- `src/yb/master/backfill_index.cc:657-668` (v2.2.0.0) - `BackfillTable::UpdateSafeTime` on a non-OK status calls
  `AlterTableStateToAbort()` (`:766`) -> `MultiStageAlterTable::UpdateIndexPermission` -> `SysCatalogTable::MutateItems`
  -> `SysCatalogTable::SyncWrite`, which blocks on a `CountDownLatch` (the issue's frames `sys_catalog.cc:627`,
  `countdown_latch.cc:66`).
- Trigger thread: `Reactor::ShutdownInternal` runs on the reactor thread (`Reactor::AsyncHandler` ->
  `ShutdownInternal`, master HEAD `src/yb/rpc/reactor.cc:524`, `:541-543`) and aborts every scheduled `DelayedTask`
  with `AbortedError()` (`reactor.cc:383-386`; status text `Aborted (yb/rpc/reactor.cc:121): Shutdown connection`).
  `DelayedTask::AbortTask` invokes the bound functor synchronously (`src/yb/rpc/delayed_task.cc:100-116`, `:112`),
  i.e. `RetryingRpcTask::RunDelayedTask(Aborted)` (v2.2.0.0 `async_rpc_tasks.cc:326-339`; master HEAD
  `src/yb/master/async_rpc_tasks_base.cc:380-393`) -> `AbortTask` -> `AbortAndReturnPrevState` (`:250-272`) ->
  `UnregisterAsyncTask` (`:417-435`) -> `UnregisterAsyncTaskCallback` (`:434`). Reactor threads have
  `wait_allowed=false`, so the latch wait fires `CHECK(LoadTLS()->wait_allowed)`
  (`src/yb/util/thread_restrictions.cc:94-95`) in debug builds; in release builds the wait silently executes on the
  reactor (the second trace in the issue shows the resulting `Waited for 1.000s trying to join with Master_reactor`).
- Shutdown ordering still makes the reactor the first aborter: `Master::Shutdown` calls `catalog_manager_->StartShutdown()`
  then `RpcAndWebServerBase::Shutdown()` (messenger/reactors) *before* `catalog_manager_->CompleteShutdown()` ->
  `AbortAndWaitForAllTasks()` (master HEAD `src/yb/master/master.cc:425-452`: `:434` StartShutdown, `:438`
  RpcAndWebServerBase::Shutdown, `:445` CompleteShutdown; `catalog_manager.cc:2354`), so a pending
  retry `DelayedTask` is seen by `RunDelayedTask` in state `kWaiting`, exactly as observed in every R cell below.

Fix (present on every current target):

- `src/yb/master/backfill_index.cc:1606-1611` (master HEAD 52c3c2bf, identical to local checkout 46a9f7ad) -
  `GetSafeTimeForTablet::UnregisterAsyncTaskCallback()` returns immediately when `state() == kAborted`
  (`VLOG(1) << " was aborted"`), never reaching `UpdateSafeTime` (`:891`). Introduced by
  `1a030bada18981361c7255834a77d1c3a63f1531` "[#13891] DocDB: Fix Potential Deadlock on Table Delete" (D19655,
  2022-09-20), which targeted the DeleteTable deadlock but closes this path too (same abort -> callback chain).
- `src/yb/master/backfill_index.cc:1782-1786` - identical guard in `BackfillChunk::UnregisterAsyncTaskCallback()`
  (the "backfill" half of the title), introduced by `bc4a465a1f90d30466d1bbe1102553e582bef022`
  "[#8774, #8780] Fix DeleteTable interaction with backfill" (D11831, 2021-07-02).
- Remaining work on the abort chain is wait-free: `AbortIfScheduled` -> `Messenger::AbortOnReactor` ->
  `DelayedTask::MarkAsDone` returns `kAlreadyDone` (`delayed_task.cc:81-94`, `:100-109`); `Finished()` is the empty
  default for backfill tasks (`async_rpc_tasks_base.h:159`, no override in `backfill_index.h`);
  `RetryingTSRpcTaskWithTable::UnregisterAsyncTaskCallbackInternal` (`async_rpc_tasks_base.cc:644-650`) ->
  `CatalogManager::CheckTableDeleted` defers its sys-catalog write to `async_task_pool_->SubmitFunc`
  (master HEAD `catalog_manager.cc:13734-13745`; 2024.2.11.0 `:12518/:12529`, 2025.1.4.0 `:13030/:13041`,
  2025.2.6.0 `:13569/:13580`, 2026.1.1.1 `:13635/:13646`, 2.25.2.0 `:12693/:12704`).

Per-tag guard sites (function start / guard VLOG line in `src/yb/master/backfill_index.cc`; `RunDelayedTask` warning line):
2024.2.11.0 `:1436/:1438` + `:1608/:1610`, `async_rpc_tasks.cc:416/:423`; 2025.1.4.0 `:1470/:1472` + `:1642/:1644`,
`async_rpc_tasks_base.cc:378/:385`; 2025.2.6.0 `:1447/:1450` + `:1623/:1625`, `async_rpc_tasks_base.cc:378/:385`;
2026.1.1.1 `:1536/:1539` + `:1712/:1714`, `async_rpc_tasks_base.cc:378/:385`; 2.25.2.0 `:1407/:1409` + `:1578/:1580`,
`async_rpc_tasks_base.cc:378/:385`; master HEAD `:1606/:1609` + `:1782/:1784`, `async_rpc_tasks_base.cc:380/:387`.
`thread_restrictions.h:37-38` (`#ifndef NDEBUG` / `ENABLE_THREAD_RESTRICTIONS`) is identical on all five tags.

Targets still affected: none.

## Trigger matrix
Predicted from source before running (release images; "abort path" = `RunDelayedTask` warning
`Async tablet task failed or was cancelled: Aborted` emitted on the reactor thread with the task still `kWaiting`):

| condition | expected | observed |
|---|---|---|
| R: user-enforced YCQL index backfill, `GetSafeTimeForTablet` in retry backoff (tserver killed), graceful master stop | abort path fires on the reactor; guard `was aborted` at the `GetSafeTimeForTablet` callback line; no `Failed backfill. Could not compute safe time`, no `Marking ... as FAILED`, no `Backfill failed/aborted.`; exit 0 fast | all 5 targets: `reactor_abort_warn=1`, `guard_vlog=2` (same thread id as the warning), all three write-chain markers 0, `fatal=0`, `join_wait=0`, exit 0 in 109-114 ms |
| C: transactional index fully backfilled, tserver alive, graceful master stop | no abort-path lines, exit 0 fast | all 5 targets: all counters 0, exit 0 in 110-115 ms (2026.1 first run 30073 ms with 14 `trying to join with iotp_call_home_0` waits = CallHome HTTP thread join, unrelated; identical re-run 110 ms, 0 waits) |
| era code (v2.2.0.0 source, no guard) | callback continues into `UpdateSafeTime(InternalError)` -> `AlterTableStateToAbort` -> `SyncWrite` on the reactor; debug: FATAL `Check failed: LoadTLS()->wait_allowed`; release: latch wait on the reactor, reactor join stalls | not executable here (no release image can show the CHECK; a 2020 debug build is out of scope); established from the tag tree and the issue's own two traces |
| debug-build FATAL on current release images | impossible: `AssertWaitAllowed(){}` under `NDEBUG` | `fatal=0` everywhere, consistent |

## Reproduction
Script: `/tmp/claude/yb4681/yb4681-matrix.sh` (host driver with the in-container program embedded; podman, arm64 images).
Consolidated results: `/tmp/claude/yb4681/yb4681-matrix.txt`.

Per cell, inside one `docker.io/yugabytedb/yugabyte:<tag>` container (RF1, loopback):
1. `yb-master --replication_factor=1 --TEST_running_test=true --disable_index_backfill_for_non_txn_tables=false
   --vmodule=backfill_index=3,async_rpc_tasks=1,async_rpc_tasks_base=1 ...` (`TEST_running_test` opts the master into
   graceful SIGTERM handling on 2024.2/2025.1/2.25.2, `termination_monitor.cc:74`; 2025.2 also accepts it,
   2026.1 is graceful by default) and `yb-tserver --enable_ysql=false
   --index_backfill_upperbound_for_user_enforced_txn_duration_ms=65000 ...`.
2. R: `CREATE TABLE ks.t (k int PRIMARY KEY, v int)` (non-transactional), 20 rows, then in the background
   `CREATE INDEX idx_v ON ks.t (v) WITH transactions = { 'enabled' : false, 'consistency_level' : 'user_enforced' }`.
   Wait for the master log line `Send GetSafeTime for ...` (2 s on every target), wait 2 s, SIGKILL the tserver,
   wait 12 s: the master's GetSafeTime RPCs now fail fast and `RescheduleWithBackoffDelay` schedules 10 retries with
   growing delay (last two: 4109 ms, 8205 ms), so a `DelayedTask` is pending on a reactor at the stop signal.
3. C: `CREATE TABLE ... WITH transactions = { 'enabled' : true }`, 20 rows, `CREATE INDEX idx_v ON ks.t (v)`; wait
   for `Completed backfilling the index table` (1 s), wait 2 s.
4. Send SIGTERM to the master, wait for exit (60 s budget), count the marker lines emitted after the signal.

Failure signature to grep for (unfixed code, release build): `Failed backfill. Could not compute safe time` and
`Marking ... as FAILED` appearing after the stop signal on the same thread id as the
`Async tablet task failed or was cancelled: Aborted` warning; unfixed debug build:
`thread_restrictions.cc] Check failed: LoadTLS()->wait_allowed Waiting is not allowed to be used on this thread`.
Fixed signature (observed): the warning followed by two `backfill_index.cc:<guard line>] vlog1:  was aborted` lines on the
same thread and `Master@127.0.0.1:7100 shutdown complete.` within ~30 ms.

Representative R excerpt (2024.2.11.0, thread 51 = reactor):
`W ... 51 async_rpc_tasks.cc:423] GetSafeTime for <tablet> (task=..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:121): Shutdown connection (system error 108)`
`I ... 51 backfill_index.cc:1438] vlog1:  was aborted` (x2)

## Versions tested
Resolved at: 2026-09-06T15:43Z by the controller and independently re-verified at 2026-09-06T16:00Z
Resolution evidence: Docker Hub `repositories/yugabytedb/yugabyte/tags/?name=<branch>&page_size=3&ordering=last_updated`
for 2024.2 (87 tags), 2025.1 (36), 2025.2 (39), 2026.1 (12), `2.25.` (9), `2.27.`/`2.29.`/`2026.2` (0 each);
`tags/latest` = `2026.1.1.1-b2` manifest sha256:3926eedf0ff4b99426fbff533128ff16635ed5697a5cf086dde4fafec28e6f75;
per-image `podman image inspect` digests and in-image `yb-master --version` (revision + `build_type RELEASE`) match;
tag SHAs from the local git tags; master HEAD from `GET repos/yugabyte/yugabyte-db/branches/master` =
52c3c2bf5800aad3e9ae74f4d41b224868caac09 (2026-09-05T19:58:56Z). Zero drift from the controller map.

| target | latest resolved version | source revision | observed repro (R) / control (C) | effective status |
|---|---|---|---|---|
| 2024.2 | 2024.2.11.0-b36 (`2024.2.11.0-b36-aarch64`, sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288, pushed 2026-08-25T15:59Z) | v2024.2.11.0 = 04544fd94fb4e50d05994784f444b075a4a5dfdb | R: abort path fired (`async_rpc_tasks.cc:423`), guard x2 at `backfill_index.cc:1438`, no write-chain markers, exit 0 in 111 ms / C: clean, exit 0 in 111 ms | fixed |
| 2025.1 | 2025.1.4.0-b103 (`...-aarch64`, sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0, pushed 2026-04-20T15:57Z) | v2025.1.4.0 = 690ac4141e57d060fd834312f396643e8c41f260 | R: abort path fired (`async_rpc_tasks_base.cc:385`), guard x2 at `backfill_index.cc:1472`, no markers, exit 0 in 113 ms / C: clean, exit 0 in 112 ms | fixed |
| 2025.2 | 2025.2.6.0-b111 (`...-aarch64`, sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e, pushed 2026-09-04T22:22Z) | v2025.2.6.0 = ae22ef35ff65de63303467d95ac3e6d2382feba9 | R: abort path fired, guard x2 at `backfill_index.cc:1450`, no markers, exit 0 in 111 ms / C: clean, exit 0 in 115 ms | fixed |
| 2026.1 | 2026.1.1.1-b2 = `latest` (`...-aarch64`, sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699, pushed 2026-08-18T20:01Z) | v2026.1.1.1 = 3b010e021f6027744bf9e016b17e108a2eb24e44 | R: abort path fired, guard x2 at `backfill_index.cc:1539`, no markers, exit 0 in 113 ms (dry run 109 ms) / C: clean, exit 0; first run 30073 ms held only by the `iotp_call_home_0` join (unrelated), identical re-run 110 ms | fixed |
| master | preview 2.25.2.0-b359 (`...-aarch64`, sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5, pushed 2025-05-20T18:32Z) | in-image 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d; master HEAD 52c3c2bf5800aad3e9ae74f4d41b224868caac09 (source-only: guards at `backfill_index.cc:1609`/`:1784`, `async_rpc_tasks_base.cc:380-393`) | R: abort path fired, guard x2 at `backfill_index.cc:1409`, no markers, exit 0 in 114 ms / C: clean, exit 0 in 110 ms | fixed (runtime on the preview; HEAD by source) |
| reported | 2020-06-04 master debug build (pre-v2.2.0.0) | v2.2.0.0 = a5f0657dcf4db948dd0c4e336ee05dd15b06869b used as the era source | not runnable (debug-only CHECK; no debug image); source shows the unguarded callback at `backfill_index.cc:1030-1049` | affected (historical) |

First-fixed releases (provenance from tag content + GitHub compare ancestry, not runtime proof):
- BackfillChunk guard `bc4a465a` (D11831): first tag v2.7.2 (2021-07-23; v2.7.1.1 lacks it), first stable v2.8.0; not
  backported to 2.4/2.6 (v2.6.20.0 lacks it).
- GetSafeTimeForTablet guard `1a030bad` (D19655): first tag v2.15.3.0 (2022-10-21; v2.15.2.1 lacks it), v2.16.0.0
  onward; 2.14 backport `38373a3f238c7c8f09c8c20460aedfb08d5f40b3` (D20391) first in v2.14.5.0 (2022-11-08; v2.14.4.0
  lacks it); not in 2.12/2.13.
- Both guards are ancestors of v2024.2.11.0, v2025.1.4.0, v2025.2.6.0, v2026.1.1.1 and v2.25.2.0 (`compare` status
  `ahead`) and present in the first tag of every in-support branch (v2024.2.0.0, v2025.1.0.0, v2025.2.0.0, v2026.1.0.0).
- No commit references #4681 or DB-5947; the issue was fixed as a side effect of #8774/#8780 and #13891.

## Proposed fix direction
No product change required; recommend closing #4681 as fixed by `bc4a465a` (#8774/#8780) and `1a030bad` (#13891),
noting that both guards are `kAborted` early returns in the backfill task callbacks. Optional hardening if the
maintainers want a structural guarantee rather than per-callback guards:
- `src/yb/master/async_rpc_tasks_base.cc` `RetryingRpcTask::RunDelayedTask`: when invoked with a non-OK status
  (reactor abort), hop `AbortTask`/`UnregisterAsyncTask` onto `callback_pool_` (as `RpcCallback` already does for
  responses, `:283-294`) so no subclass callback can ever run on a reactor thread; keep the direct call only when
  the pool submit fails.
- Alternatively add `DCHECK(ThreadRestrictions::IsWaitAllowed())` at the top of `SysCatalogTable::SyncWrite` callers
  reachable from task callbacks, so any future regression surfaces in debug tests with a direct frame.
Edge cases: `UnregisterAsyncTask` is called twice on this path (`:389` then `:391`); the guard must stay idempotent.

## Test strategy
- Existing coverage: `yb-admin-test_ent` `AdminCliTest.TestExportImportIndexSnapshot` (the reporter's test), and the
  fix commits' plans `org.yb.cql.TestPartialIndex#testUniquePartialIndexPart*` (#13891) and
  `org.yb.cql.TestTransaction#testTransactionConditionalDMLWithElseError` (#8774).
- Deterministic debug-build check (public entry points only): mini-cluster test that creates a user-enforced YCQL
  index with `--disable_index_backfill_for_non_txn_tables=false`, sets tserver
  `--index_backfill_upperbound_for_user_enforced_txn_duration_ms=65000`, kills the tserver once
  `Send GetSafeTime for` is logged, waits for `Scheduling retry with a delay of` >= 4 s, then shuts the master down;
  assert no `Failed backfill. Could not compute safe time` after shutdown start and that the master joins its
  reactors without `trying to join with Master_reactor`. In debug builds the ThreadRestrictions CHECK is the oracle.
- The container script above is the release-image equivalent (`bash /tmp/claude/yb4681/yb4681-matrix.sh <tag> R`).

## Confidentiality check
Artifact, script and matrix contain only public issue text, public commit SHAs/Phorge IDs, official Docker Hub
tags/digests, loopback addresses, synthetic schema (`ks.t`, `idx_v`) and no customer data, PII, credentials or
internal hostnames.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4681-matrix.sh.txt`):

#!/bin/bash
# yugabyte/yugabyte-db#4681 (DB-5947) runtime proxy matrix.
#
# The reported failure is a debug-only CHECK (ThreadRestrictions::AssertWaitAllowed is compiled
# out under NDEBUG, src/yb/util/thread_restrictions.h:37-39 / :123-131), so official RELEASE images
# can never print it. This probe exercises the *reported path* on release images and observes what
# the master does instead:
#
#   cell R (repro shape): a YCQL user-enforced index backfill whose GetSafeTimeForTablet task is in
#          retry backoff (tserver killed) -> a DelayedTask for RetryingRpcTask::RunDelayedTask is
#          pending on a master reactor; the master then receives a graceful stop signal (SIGTERM).
#          Reactor::ShutdownInternal aborts the DelayedTask on the reactor thread ->
#          RunDelayedTask(Aborted) -> AbortTask -> UnregisterAsyncTask -> UnregisterAsyncTaskCallback.
#          Unfixed code (v2.2.0.0 backfill_index.cc:1030-1049) continues into
#          BackfillTable::UpdateSafeTime -> AlterTableStateToAbort -> sys-catalog SyncWrite on the
#          reactor thread. Fixed code returns at the kAborted guard ("was aborted" VLOG(1)).
#   cell C (control): transactional index fully backfilled, tserver alive, same graceful stop.
#
# Signals recorded per cell (master log lines emitted AFTER the stop signal):
#   reactor_abort_warn  = "Async tablet task failed or was cancelled" (RunDelayedTask invoked with a
#                         non-OK status by the reactor abort -> the reported trigger fired)
#   guard_vlog          = " was aborted" (kAborted early return in the backfill task callbacks)
#   updatesafetime_err  = "Failed backfill. Could not compute safe time" (UpdateSafeTime error branch;
#                         the first step of the reported sys-catalog write chain)
#   mark_failed         = "Marking ... as FAILED" (MarkIndexesAsDesired -> sys_catalog Upsert)
#   backfill_failed     = "Backfill failed/aborted." (BackfillTable::Abort)
#   fatal               = FATAL / "Check failed" / "Waiting is not allowed"
#   join_wait           = "trying to join with" (thread join stall, second trace in the issue)
#
# Usage: bash /tmp/claude/yb4681/yb4681-matrix.sh            # all five targets, cells R and C
#        bash /tmp/claude/yb4681/yb4681-matrix.sh 2026.1.1.1-b2-aarch64 R   # one cell
# Requires: podman, the five docker.io/yugabytedb/yugabyte:<tag> images (pulled beforehand).
set -u
OUT=/tmp/claude/yb4681/yb4681-matrix.txt
CELLDIR=/tmp/claude/yb4681/cells
mkdir -p "$CELLDIR"
TARGETS="2024.2.11.0-b36-aarch64 2025.1.4.0-b103-aarch64 2025.2.6.0-b111-aarch64 2026.1.1.1-b2-aarch64 2.25.2.0-b359-aarch64"
CELLS="R C"
[ $# -ge 1 ] && TARGETS="$1"
[ $# -ge 2 ] && CELLS="$2"

read -r -d '' INCONTAINER <<'EOS'
set -u
BIN=/home/yugabyte/bin
D=/tmp/yb4681; rm -rf $D; mkdir -p $D/m $D/t
ts() { date -u +%H:%M:%S.%N | cut -c1-12; }
echo "VERSION: $($BIN/yb-master --version 2>&1 | head -1)"
echo "CELL: $CELL start=$(ts)"
$BIN/yb-master --fs_data_dirs=$D/m --master_addresses=127.0.0.1:7100 --rpc_bind_addresses=127.0.0.1:7100 \
  --webserver_interface=127.0.0.1 --webserver_port=7000 --replication_factor=1 \
  --logtostderr --vmodule=backfill_index=3,async_rpc_tasks=1,async_rpc_tasks_base=1 \
  --TEST_running_test=true --disable_index_backfill_for_non_txn_tables=false \
  > $D/master.log 2>&1 &
MPID=$!
$BIN/yb-tserver --fs_data_dirs=$D/t --tserver_master_addrs=127.0.0.1:7100 --rpc_bind_addresses=127.0.0.1:9100 \
  --webserver_interface=127.0.0.1 --webserver_port=9000 --cql_proxy_bind_address=127.0.0.1:9042 \
  --enable_ysql=false --logtostderr \
  --index_backfill_upperbound_for_user_enforced_txn_duration_ms=65000 \
  > $D/tserver.log 2>&1 &
TPID=$!
ok=0; for i in $(seq 1 150); do if $BIN/ycqlsh 127.0.0.1 9042 -e 'SELECT now() FROM system.local;' >/dev/null 2>&1; then ok=1; break; fi; sleep 1; done
echo "CQL_READY: $ok after ${i}s at $(ts)"
if [ $ok != 1 ]; then echo "RESULT cell=$CELL exit=NA reason=cql_not_ready"; tail -5 $D/master.log; tail -5 $D/tserver.log; exit 2; fi
if [ "$CELL" = R ]; then
  $BIN/ycqlsh 127.0.0.1 9042 -e "CREATE KEYSPACE ks; CREATE TABLE ks.t (k int PRIMARY KEY, v int);"
  for k in $(seq 1 20); do echo "INSERT INTO ks.t (k, v) VALUES ($k, $((k % 5)));"; done | $BIN/ycqlsh 127.0.0.1 9042
  ( timeout 300 $BIN/ycqlsh 127.0.0.1 9042 --request-timeout=300 -e "CREATE INDEX idx_v ON ks.t (v) WITH transactions = { 'enabled' : false, 'consistency_level' : 'user_enforced' };" > $D/create_index.out 2>&1 ) &
  got=0; for i in $(seq 1 90); do if grep -q 'Send GetSafeTime for' $D/master.log; then got=1; break; fi; sleep 1; done
  echo "GETSAFETIME_SENT: $got after ${i}s at $(ts)"
  sleep 2
  echo "KILL_TSERVER at $(ts)"; kill -9 $TPID; wait $TPID 2>/dev/null
  sleep 12
  echo "RETRY_SCHEDULES_BEFORE_STOP: $(grep -c 'Scheduling retry with a delay of' $D/master.log)"
  grep 'Scheduling retry with a delay of' $D/master.log | tail -2 | sed -E 's/^(.{15}).*(GetSafeTime for [0-9a-f]{8})[^ ]* .*(Scheduling retry with a delay of [0-9]+ms).*/\1 \2... \3/'
else
  $BIN/ycqlsh 127.0.0.1 9042 -e "CREATE KEYSPACE ks; CREATE TABLE ks.t (k int PRIMARY KEY, v int) WITH transactions = { 'enabled' : true };"
  for k in $(seq 1 20); do echo "INSERT INTO ks.t (k, v) VALUES ($k, $((k % 5)));"; done | $BIN/ycqlsh 127.0.0.1 9042
  timeout 120 $BIN/ycqlsh 127.0.0.1 9042 --request-timeout=120 -e "CREATE INDEX idx_v ON ks.t (v);" > $D/create_index.out 2>&1
  got=0; for i in $(seq 1 90); do if grep -q 'Completed backfilling the index table' $D/master.log; then got=1; break; fi; sleep 1; done
  echo "BACKFILL_COMPLETED: $got after ${i}s at $(ts)"
  sleep 2
fi
N0=$(wc -l < $D/master.log)
echo "STOP_SIGNAL_MASTER at $(ts) (master log lines before stop: $N0)"
T0=$(date +%s%N)
kill -TERM $MPID
alive=1; for i in $(seq 1 600); do if ! kill -0 $MPID 2>/dev/null; then alive=0; break; fi; sleep 0.1; done
T1=$(date +%s%N)
if [ $alive = 1 ]; then echo "MASTER_STILL_ALIVE after 60s; sending SIGKILL"; kill -9 $MPID; EC=KILLED; else wait $MPID; EC=$?; fi
MS=$(( (T1 - T0) / 1000000 ))
tail -n +$((N0+1)) $D/master.log > $D/after.log
c() { grep -c -- "$1" $D/after.log; }
echo "RESULT cell=$CELL exit=$EC stop_to_exit_ms=$MS reactor_abort_warn=$(c 'Async tablet task failed or was cancelled') guard_vlog=$(c 'was aborted') updatesafetime_err=$(c 'Failed backfill. Could not compute safe time') mark_failed=$(grep -c 'Marking .* as FAILED' $D/after.log) backfill_failed=$(c 'Backfill failed/aborted') fatal=$(grep -c -E '^F[0-9]|Check failed|Waiting is not allowed' $D/after.log) join_wait=$(c 'trying to join with') shutdown_complete=$(c 'shutdown complete')"
echo "GUARD_SITES: $(grep 'was aborted' $D/after.log | sed -E 's/.*(backfill_index.cc:[0-9]+).*/\1/' | sort | uniq -c | tr '\n' ';')"
echo "KEY_LINES_AFTER_STOP:"
grep -E 'Async tablet task failed or was cancelled|was aborted|Failed backfill|Marking .* as FAILED|Backfill failed/aborted|trying to join with|shutdown complete|shutting down\.\.\.' $D/after.log | sed -E 's/Backfilling index tables \{ idx_v \} //; s/0x0000[0-9a-f]{12}/0x.../' | cut -c1-330 | head -12
EOS

run_cell() {
  local tag="$1" cell="$2"
  local name="yb4681-${cell}-$(echo "$tag" | tr '.' '_')"
  local f="$CELLDIR/${tag}-${cell}.txt"
  podman rm -f "$name" >/dev/null 2>&1
  {
    echo "===== target=$tag cell=$cell image=docker.io/yugabytedb/yugabyte:$tag ====="
    echo "IMAGE_DIGEST: $(podman image inspect docker.io/yugabytedb/yugabyte:$tag --format '{{.Digest}}')"
    # macOS has no coreutils `timeout`; perl alarm is the host-side watchdog (in-container script is bounded anyway).
    perl -e 'alarm 600; exec @ARGV' podman run --rm --name "$name" --entrypoint bash -e CELL="$cell" \
      "docker.io/yugabytedb/yugabyte:$tag" -c "$INCONTAINER" 2>&1
    echo "PODMAN_EXIT: $?"
  } > "$f"
  cat "$f"
}

for cell in $CELLS; do
  for tag in $TARGETS; do
    run_cell "$tag" "$cell"
  done
done

# Consolidate into ONE evidence file (only when the full default matrix was requested).
if [ $# -eq 0 ]; then
  {
    echo "yugabyte/yugabyte-db#4681 (DB-5947) runtime proxy matrix -- generated by yb4681-matrix.sh"
    echo "Host: podman $(podman version --format '{{.Client.Version}}') on Apple Silicon (arm64 images)"
    echo "Note: release images compile out ThreadRestrictions::AssertWaitAllowed (NDEBUG); the FATAL in the"
    echo "issue cannot appear here. Cells show whether the reported abort path fires and how it is handled."
    echo
    for cell in R C; do for tag in $TARGETS; do cat "$CELLDIR/${tag}-${cell}.txt"; echo; done; done
  } > "$OUT"
  echo "Wrote $OUT"
fi

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4681-matrix.txt`):

yugabyte/yugabyte-db#4681 (DB-5947) runtime proxy matrix -- generated by yb4681-matrix.sh
Host: podman 5.8.6 on Apple Silicon (arm64 images)
Note: release images compile out ThreadRestrictions::AssertWaitAllowed (NDEBUG); the FATAL in the
issue cannot appear here. Cells show whether the reported abort path fires and how it is handled.

===== target=2024.2.11.0-b36-aarch64 cell=R image=docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64 =====
IMAGE_DIGEST: sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288
VERSION: version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
CELL: R start=16:05:35.818
CQL_READY: 1 after 5s at 16:05:41.694
GETSAFETIME_SENT: 1 after 2s at 16:05:43.567
KILL_TSERVER at 16:05:45.575
RETRY_SCHEDULES_BEFORE_STOP: 10
I0906 16:05:49. GetSafeTime for fa31d266... Scheduling retry with a delay of 4109ms
I0906 16:05:53. GetSafeTime for fa31d266... Scheduling retry with a delay of 8205ms
STOP_SIGNAL_MASTER at 16:05:57.600 (master log lines before stop: 2342)
RESULT cell=R exit=0 stop_to_exit_ms=111 reactor_abort_warn=1 guard_vlog=2 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES:       2 backfill_index.cc:1438;
KEY_LINES_AFTER_STOP:
I0906 16:05:57.614400    10 master.cc:404] Master@127.0.0.1:7100 shutting down...
W0906 16:05:57.625948    51 async_rpc_tasks.cc:423] GetSafeTime for fa31d266a96c4a5e8b1dded0ceedf0e3 (task=0x..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:121): Shutdown connection (system error 108)
I0906 16:05:57.626199    51 backfill_index.cc:1438] vlog1:  was aborted
I0906 16:05:57.626406    51 backfill_index.cc:1438] vlog1:  was aborted
I0906 16:05:57.654011    10 master.cc:422] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2025.1.4.0-b103-aarch64 cell=R image=docker.io/yugabytedb/yugabyte:2025.1.4.0-b103-aarch64 =====
IMAGE_DIGEST: sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0
VERSION: version 2025.1.4.0 build 103 revision 690ac4141e57d060fd834312f396643e8c41f260 build_type RELEASE built at 02 Apr 2026 20:43:14 UTC
CELL: R start=16:06:06.631
CQL_READY: 1 after 5s at 16:06:12.143
GETSAFETIME_SENT: 1 after 2s at 16:06:13.835
KILL_TSERVER at 16:06:15.847
RETRY_SCHEDULES_BEFORE_STOP: 10
I0906 16:06:20. GetSafeTime for 60eb7347... Scheduling retry with a delay of 4109ms
I0906 16:06:24. GetSafeTime for 60eb7347... Scheduling retry with a delay of 8205ms
STOP_SIGNAL_MASTER at 16:06:27.876 (master log lines before stop: 2818)
RESULT cell=R exit=0 stop_to_exit_ms=113 reactor_abort_warn=1 guard_vlog=2 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES:       2 backfill_index.cc:1472;
KEY_LINES_AFTER_STOP:
I0906 16:06:27.892023    10 master.cc:418] Master@127.0.0.1:7100 shutting down...
W0906 16:06:27.903931    48 async_rpc_tasks_base.cc:385] GetSafeTime for 60eb73477a8d4d6992a5e780410e6dd4 (task=0x..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:121): Shutdown connection (system error 108)
I0906 16:06:27.904840    48 backfill_index.cc:1472] vlog1:  was aborted
I0906 16:06:27.905179    48 backfill_index.cc:1472] vlog1:  was aborted
I0906 16:06:27.927434    10 master.cc:436] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2025.2.6.0-b111-aarch64 cell=R image=docker.io/yugabytedb/yugabyte:2025.2.6.0-b111-aarch64 =====
IMAGE_DIGEST: sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e
VERSION: version 2025.2.6.0 build 111 revision ae22ef35ff65de63303467d95ac3e6d2382feba9 build_type RELEASE built at 02 Sep 2026 16:42:37 UTC
CELL: R start=16:06:38.472
CQL_READY: 1 after 5s at 16:06:44.443
GETSAFETIME_SENT: 1 after 2s at 16:06:46.261
KILL_TSERVER at 16:06:48.275
RETRY_SCHEDULES_BEFORE_STOP: 10
I0906 16:06:52. GetSafeTime for 762f3c77... Scheduling retry with a delay of 4109ms
I0906 16:06:56. GetSafeTime for 762f3c77... Scheduling retry with a delay of 8205ms
STOP_SIGNAL_MASTER at 16:07:00.302 (master log lines before stop: 3270)
RESULT cell=R exit=0 stop_to_exit_ms=111 reactor_abort_warn=1 guard_vlog=2 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES:       2 backfill_index.cc:1450;
KEY_LINES_AFTER_STOP:
I0906 16:07:00.317200    10 master.cc:421] Master@127.0.0.1:7100 shutting down...
W0906 16:07:00.324219    49 async_rpc_tasks_base.cc:385] GetSafeTime for 762f3c77c7824fccbb22fb31afed7e44 (task=0x..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:121): Shutdown connection (system error 108)
I0906 16:07:00.324550    49 backfill_index.cc:1450] vlog1:  was aborted
I0906 16:07:00.324663    49 backfill_index.cc:1450] vlog1:  was aborted
I0906 16:07:00.339816    10 master.cc:439] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2026.1.1.1-b2-aarch64 cell=R image=docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 =====
IMAGE_DIGEST: sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699
VERSION: version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
CELL: R start=16:07:10.541
CQL_READY: 1 after 5s at 16:07:16.136
GETSAFETIME_SENT: 1 after 2s at 16:07:17.757
KILL_TSERVER at 16:07:19.769
RETRY_SCHEDULES_BEFORE_STOP: 10
I0906 16:07:23. GetSafeTime for 90693340... Scheduling retry with a delay of 4109ms
I0906 16:07:28. GetSafeTime for 90693340... Scheduling retry with a delay of 8205ms
STOP_SIGNAL_MASTER at 16:07:31.797 (master log lines before stop: 3082)
RESULT cell=R exit=0 stop_to_exit_ms=113 reactor_abort_warn=1 guard_vlog=2 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES:       2 backfill_index.cc:1539;
KEY_LINES_AFTER_STOP:
I0906 16:07:31.811964    10 master.cc:420] Master@127.0.0.1:7100 shutting down...
W0906 16:07:31.854091    45 async_rpc_tasks_base.cc:385] GetSafeTime for 90693340a2a043419c2c43cf105fe501 (task=0x..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:122): Shutdown connection (system error 108)
I0906 16:07:31.854393    45 backfill_index.cc:1539] vlog1:  was aborted
I0906 16:07:31.854568    45 backfill_index.cc:1539] vlog1:  was aborted
I0906 16:07:31.869092    10 master.cc:438] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2.25.2.0-b359-aarch64 cell=R image=docker.io/yugabytedb/yugabyte:2.25.2.0-b359-aarch64 =====
IMAGE_DIGEST: sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5
VERSION: version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
CELL: R start=16:07:40.656
CQL_READY: 1 after 5s at 16:07:46.260
GETSAFETIME_SENT: 1 after 2s at 16:07:47.925
KILL_TSERVER at 16:07:49.937
RETRY_SCHEDULES_BEFORE_STOP: 10
I0906 16:07:54. GetSafeTime for a00d4700... Scheduling retry with a delay of 4109ms
I0906 16:07:58. GetSafeTime for a00d4700... Scheduling retry with a delay of 8205ms
STOP_SIGNAL_MASTER at 16:08:01.972 (master log lines before stop: 2347)
RESULT cell=R exit=0 stop_to_exit_ms=114 reactor_abort_warn=1 guard_vlog=2 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES:       2 backfill_index.cc:1409;
KEY_LINES_AFTER_STOP:
I0906 16:08:01.990497    10 master.cc:407] Master@127.0.0.1:7100 shutting down...
W0906 16:08:01.991907    50 async_rpc_tasks_base.cc:385] GetSafeTime for a00d47004d1d41a998dbf3fd09b667be (task=0x..., state=kWaiting): Async tablet task failed or was cancelled: Aborted (yb/rpc/reactor.cc:121): Shutdown connection (system error 108)
I0906 16:08:01.992134    50 backfill_index.cc:1409] vlog1:  was aborted
I0906 16:08:01.992434    50 backfill_index.cc:1409] vlog1:  was aborted
I0906 16:08:02.011130    10 master.cc:425] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2024.2.11.0-b36-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64 =====
IMAGE_DIGEST: sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288
VERSION: version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
CELL: C start=16:08:12.667
CQL_READY: 1 after 5s at 16:08:18.496
BACKFILL_COMPLETED: 1 after 1s at 16:08:19.493
STOP_SIGNAL_MASTER at 16:08:21.504 (master log lines before stop: 2414)
RESULT cell=C exit=0 stop_to_exit_ms=111 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
I0906 16:08:21.519290    10 master.cc:404] Master@127.0.0.1:7100 shutting down...
I0906 16:08:21.560461    10 master.cc:422] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2025.1.4.0-b103-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2025.1.4.0-b103-aarch64 =====
IMAGE_DIGEST: sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0
VERSION: version 2025.1.4.0 build 103 revision 690ac4141e57d060fd834312f396643e8c41f260 build_type RELEASE built at 02 Apr 2026 20:43:14 UTC
CELL: C start=16:08:30.892
CQL_READY: 1 after 5s at 16:08:36.716
BACKFILL_COMPLETED: 1 after 1s at 16:08:37.874
STOP_SIGNAL_MASTER at 16:08:39.886 (master log lines before stop: 2890)
RESULT cell=C exit=0 stop_to_exit_ms=112 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
I0906 16:08:39.902650    10 master.cc:418] Master@127.0.0.1:7100 shutting down...
I0906 16:08:39.941972    10 master.cc:436] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2025.2.6.0-b111-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2025.2.6.0-b111-aarch64 =====
IMAGE_DIGEST: sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e
VERSION: version 2025.2.6.0 build 111 revision ae22ef35ff65de63303467d95ac3e6d2382feba9 build_type RELEASE built at 02 Sep 2026 16:42:37 UTC
CELL: C start=16:08:51.608
CQL_READY: 1 after 5s at 16:08:57.531
BACKFILL_COMPLETED: 1 after 1s at 16:08:58.582
STOP_SIGNAL_MASTER at 16:09:00.596 (master log lines before stop: 3140)
RESULT cell=C exit=0 stop_to_exit_ms=115 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
I0906 16:09:00.612953    10 master.cc:421] Master@127.0.0.1:7100 shutting down...
I0906 16:09:00.654242    10 master.cc:439] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

===== target=2026.1.1.1-b2-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 =====
IMAGE_DIGEST: sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699
VERSION: version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
CELL: C start=16:09:13.532
CQL_READY: 1 after 6s at 16:09:20.988
BACKFILL_COMPLETED: 1 after 1s at 16:09:22.214
STOP_SIGNAL_MASTER at 16:09:24.229 (master log lines before stop: 3153)
RESULT cell=C exit=0 stop_to_exit_ms=30073 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=14 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
W0906 16:09:40.236001    10 thread.cc:831] Waited for 1.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:41.236281    10 thread.cc:831] Waited for 2.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:42.237160    10 thread.cc:831] Waited for 3.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:43.238348    10 thread.cc:831] Waited for 4.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:44.239687    10 thread.cc:831] Waited for 5.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:45.245944    10 thread.cc:831] Waited for 6.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:46.246317    10 thread.cc:831] Waited for 7.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:47.246832    10 thread.cc:831] Waited for 8.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:48.246992    10 thread.cc:831] Waited for 9.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:49.247222    10 thread.cc:831] Waited for 10.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:50.249210    10 thread.cc:831] Waited for 11.000s trying to join with iotp_call_home_0 (tid 76)
W0906 16:09:51.249776    10 thread.cc:831] Waited for 12.000s trying to join with iotp_call_home_0 (tid 76)
PODMAN_EXIT: 0

===== target=2.25.2.0-b359-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2.25.2.0-b359-aarch64 =====
IMAGE_DIGEST: sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5
VERSION: version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
CELL: C start=16:10:02.508
CQL_READY: 1 after 5s at 16:10:08.042
BACKFILL_COMPLETED: 1 after 1s at 16:10:09.071
STOP_SIGNAL_MASTER at 16:10:11.085 (master log lines before stop: 2417)
RESULT cell=C exit=0 stop_to_exit_ms=110 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
I0906 16:10:11.099762    10 master.cc:407] Master@127.0.0.1:7100 shutting down...
I0906 16:10:11.140096    10 master.cc:425] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0


===== ADDENDUM: re-run of target=2026.1.1.1-b2-aarch64 cell=C (first run above exited after 30073 ms with 14 'Waited for N.000s trying to join with iotp_call_home_0' lines: the CallHome outbound HTTP thread was mid-request when the stop signal arrived and Master::Shutdown -> call_home->Shutdown() joins it with a 30 s give-up; unrelated to backfill tasks or reactor DelayedTasks; the R cell on the same image exited in 113 ms). Second run, identical command: =====
===== target=2026.1.1.1-b2-aarch64 cell=C image=docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 =====
IMAGE_DIGEST: sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699
VERSION: version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
CELL: C start=16:10:57.960
CQL_READY: 1 after 5s at 16:11:03.510
BACKFILL_COMPLETED: 1 after 1s at 16:11:04.499
STOP_SIGNAL_MASTER at 16:11:06.512 (master log lines before stop: 3146)
RESULT cell=C exit=0 stop_to_exit_ms=110 reactor_abort_warn=0 guard_vlog=0 updatesafetime_err=0 mark_failed=0 backfill_failed=0 fatal=0 join_wait=0 shutdown_complete=1
GUARD_SITES: 
KEY_LINES_AFTER_STOP:
I0906 16:11:06.526983    10 master.cc:420] Master@127.0.0.1:7100 shutting down...
I0906 16:11:06.597618    10 master.cc:438] Master@127.0.0.1:7100 shutdown complete.
PODMAN_EXIT: 0

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
